### PR TITLE
Add creator URL, creator base URL and creator origin to browsing context

### LINKED ISSUE #24944: Non-http iframes get a SyntaxError trying to use XHR
WPT xhr/open-url-javascript-window-2 and xhr/open-url-about-blank-window respectively make a javascript: iframe where the src script uses XHR, and an about:blank iframe that has new .XMLHttpRequest called on its proxy. In both cases, the test output complains of a line-numberless exception matching the SyntaxError string defined in script/dom/domexceptions.rs.

### PR DESCRIPTION
**Pull request by utsavoza**
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #24944 and fix (partially) #26287
- [x] There are tests for these changes

### CODE REVIEW

**Comment by jdm** [comment]
```
  ▶ Unexpected subtest result in /html/infrastructure/urls/terminology-0/document-base-url.html:
  └ PASS [expected FAIL] The fallback base URL of a document whose address is about:blank is the document base URL of the creator document.
  ▶ Unexpected subtest result in /html/infrastructure/urls/terminology-0/document-base-url.html:
  └ PASS [expected FAIL] about:blank with a base element.
  ▶ Unexpected subtest result in /html/webappapis/dynamic-markup-insertion/opening-the-input-stream/abort-while-navigating.window.html:
  │ TIMEOUT [expected FAIL] document.open() aborts documents that are navigating through iframe loading (XMLHttpRequest)
  └   → Test timed out
  ▶ Unexpected subtest result in /html/webappapis/dynamic-markup-insertion/opening-the-input-stream/abort-while-navigating.window.html:
  │ TIMEOUT [expected PASS] document.open() aborts documents that are navigating through iframe loading (fetch())
  └   → Test timed out
  ▶ Unexpected subtest result in /html/browsers/browsing-the-web/navigating-across-documents/010.html:
  │ FAIL [expected PASS] Link with onclick form submit to javascript url with delayed document.write and href navigation 
  │   → assert_equals: expected "href" but got "write"
  │ 
  │ onmessage<@http://web-platform.test:8000/html/browsers/browsing-the-web/navigating-across-documents/010.html:14:18
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1975:25
  └ Test.prototype.step_func/<@http://web-platform.test:8000/resources/testharness.js:2000:35
  ▶ Unexpected subtest result in /html/semantics/document-metadata/the-base-element/base_about_blank.html:
  └ PASS [expected FAIL] base element in about:blank document should resolve against its fallback base URI
```

**Comment by utsavoza** [comment]
It seems we have a couple of unexpected test fixes and failures. After sorting them out:

- These are valid test passes that I missed.
    ```
    ▶ Unexpected subtest result in /html/infrastructure/urls/terminology-0/document-base-url.html:
    └ PASS [expected FAIL] The fallback base URL of a document whose address is about:blank is     the document base URL of the creator document.

    ▶ Unexpected subtest result in /html/infrastructure/urls/terminology-0/document-base-url.html:
    └ PASS [expected FAIL] about:blank with a base element.

    ▶ Unexpected subtest result in /html/semantics/document-metadata/the-base-element/base_about_blank.html:
    └ PASS [expected FAIL] base element in about:blank document should resolve against its fallback base URI
    ```

- Gleaning from the test, I think the reason for the timeouts might be #21937. 
    ```
    ▶ Unexpected subtest result in /html/webappapis/dynamic-markup-insertion/opening-the-input-stream/abort-while-navigating.window.html:
    │ TIMEOUT [expected FAIL] document.open() aborts documents that are navigating through iframe loading (XMLHttpRequest)
    └   → Test timed out

    ▶ Unexpected subtest result in /html/webappapis/dynamic-markup-insertion/opening-the-input-stream/abort-while-navigating.window.html:
    │ TIMEOUT [expected PASS] document.open() aborts documents that are navigating through iframe loading (fetch())
    └   → Test timed out

    ```
    Specifically,
https://github.com/servo/servo/blob/0fc2155a02ed20490b9c99a1bec2949d859e61ee/components/script/dom/document.rs#L4651-L4657

- I am not really sure why this test failed. My best guess would be #25065, unless otherwise?
    ```
    ▶ Unexpected subtest result in /html/browsers/browsing-the-web/navigating-across-documents/010.html:
  │ FAIL [expected PASS] Link with onclick form submit to javascript url with delayed document.write and href navigation
  │   → assert_equals: expected "href" but got "write"
  │
  │ onmessage<@http://web-platform.test:8000/html/browsers/browsing-the-web/navigating-across-documents/010.html:14:18
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1975:25
  └ Test.prototype.step_func/<@http://web-platform.test:8000/resources/testharness.js:2000:35
    ```

**Comment by utsavoza** [comment]
Also, after going through the [follow hyperlink](https://html.spec.whatwg.org/multipage/#following-hyperlinks-2) implementation, I think we are incorrectly parsing the subject's href attribute relative to node's document. 

https://github.com/servo/servo/blob/0fc2155a02ed20490b9c99a1bec2949d859e61ee/components/script/dom/htmlanchorelement.rs#L640-L643

According to the [Parse a URL](https://html.spec.whatwg.org/multipage/urls-and-fetching.html#parse-a-url) algorithm, we should be parsing the href attribute value against the document's base URL.

File: components/script/dom/document.rs [hunk 1]
```diff
@@ -575,10 +575,19 @@ impl Document {
 
     // https://html.spec.whatwg.org/multipage/#fallback-base-url
     pub fn fallback_base_url(&self) -> ServoUrl {
-        // Step 1: iframe srcdoc (#4767).
-        // Step 2: about:blank with a creator browsing context.
-        // Step 3.
-        self.url()
+        let document_url = self.url();
+        if let Some(browsing_context) = self.browsing_context() {
+            // Step 1: If document is an iframe srcdoc document, then return the
```
Comment (jdm) [inline components/script/dom/document.rs:580]: We support srcdoc iframes, so we could get `browsing_context.parent().and_then(|parent| parent.document()).map(|doc| doc.base_url())` and implement this missing step.

File: components/script/dom/windowproxy.rs [hunk 2]
```diff
@@ -655,6 +703,56 @@ impl WindowProxy {
...
+        parent: Option<&WindowProxy>,
+        opener: Option<&WindowProxy>,
+    ) -> CreatorBrowsingContextInfo {
+        let creator = if parent.is_some() {
+            parent.unwrap().document()
+        } else if opener.is_some() {
+            opener.unwrap().document()
+        } else {
+            None
+        };
```
Comment (jdm) [inline components/script/dom/windowproxy.rs:740]: Let's write this as:
```rust
let creator = match (parent, opener) {
    (Some(parent), _) => parent.document(),
    (None, Some(opener)) => opener.document(),
    (None, None) => None,
};
```

**Review by jdm — CHANGES_REQUESTED** [review]
Looking good! This is a nice correctness improvement.

File: components/script/dom/document.rs (hunk 1 above)
Comment (utsavoza) [reply components/script/dom/document.rs:580]: Does having a nested browsing context correctly identify if the document is an iframe srcdoc here?

Comment (jdm) [reply components/script/dom/document.rs:580]: We'll want to check for an `about:srcdoc` url as well, given https://github.com/servo/servo/blob/master/components/script/script_thread.rs#L3923.

File: components/script/dom/windowproxy.rs (hunk 2 above)
Comment (utsavoza) [reply components/script/dom/windowproxy.rs]: Done

File: components/script/dom/document.rs (hunk 1 above)
Comment (utsavoza) [reply components/script/dom/document.rs:580]: Right. I have updated the changes accordingly.

**Comment by utsavoza** [comment]
Some more unexpected test failures. Looking at the changes, I think adding a check for srcdoc in the document fallback url implementation might be causing these additional test failures afaict. Will have to see what each failure entails. Let me know what you think @jdm.

**Comment by jdm** [comment]
Yes, I was expecting that outcome :)

**Comment by utsavoza** [comment]
1.  /FileAPI/url/sandboxed-iframe.html
    ```
    ▶ Unexpected subtest result in /FileAPI/url/sandboxed-iframe.html:
    │ TIMEOUT [expected PASS] Blob URLs can be used in <script> tags
    └   → Test timed out
    ▶ Unexpected subtest result in /FileAPI/url/sandboxed-iframe.html:
    │ FAIL [expected PASS] Blob URLs can be used in iframes, and are treated same origin
    └   → The operation is insecure.
    ▶ Unexpected subtest result in /FileAPI/url/sandboxed-iframe.html:
    │ TIMEOUT [expected PASS] Blob URL fragment is implemented.
    └   → Test timed out
    ▶ Unexpected subtest result in /FileAPI/url/sandboxed-iframe.html:
    │ FAIL [expected PASS] Blob URLs can be used in XHR
    └   → promise_test: Unhandled rejection with value: "Got unexpected error event"
    ▶ Unexpected subtest result in /FileAPI/url/sandboxed-iframe.html:
    │ FAIL [expected PASS] XHR with a fragment should succeed
    └   → promise_test: Unhandled rejection with value: "Got unexpected error event"
    ▶ Unexpected subtest result in /FileAPI/url/sandboxed-iframe.html:
    │ FAIL [expected PASS] Only exact matches should revoke URLs, using XHR
    └   → promise_test: Unhandled rejection with value: "Got unexpected error event"
    ▶ Unexpected subtest result in /FileAPI/url/sandboxed-iframe.html:
    │ FAIL [expected PASS] Revoke blob URL after open(), will fetch
    └   → assert_unreached: Got unexpected error event Reached unreachable code
    ▶ Unexpected subtest result in /FileAPI/url/sandboxed-iframe.html:
    │ FAIL [expected PASS] Blob URLs can be used in fetch
    └   → promise_test: Unhandled rejection with value: object "TypeError: Network error occurred"
    ▶ Unexpected subtest result in /FileAPI/url/sandboxed-iframe.html:
    │ FAIL [expected PASS] fetch with a fragment should succeed
    └   → promise_test: Unhandled rejection with value: object "TypeError: Network error occurred"
    ▶ Unexpected subtest result in /FileAPI/url/sandboxed-iframe.html:
    │ FAIL [expected PASS] Only exact matches should revoke URLs, using fetch
    └   → promise_test: Unhandled rejection with value: object "TypeError: Network error occurred"
    ▶ Unexpected subtest result in /FileAPI/url/sandboxed-iframe.html:
    │ FAIL [expected PASS] fetch should return Content-Type from Blob
    └   → promise_test: Unhandled rejection with value: object "TypeError: Network error occurred"
    ▶ Unexpected subtest result in /FileAPI/url/sandboxed-iframe.html:
    │ FAIL [expected PASS] Revoke blob URL after creating Request, will fetch
    └   → promise_test: Unhandled rejection with value: object "TypeError: Network error occurred"
    ▶ Unexpected subtest result in /FileAPI/url/sandboxed-iframe.html:
    │ FAIL [expected PASS] Revoke blob URL after calling fetch, fetch should succeed
    └   → promise_test: Unhandled rejection with value: object "TypeError: Network error occurred"
    ▶ Unexpected subtest result in /FileAPI/url/sandboxed-iframe.html:
    │ FAIL [expected PASS] Blob URL parses correctly
    │   → assert_true: Path must end with a valid UUID expected true got false
    │
    │ @http://web-platform.test:8000/FileAPI/url/url-format.any.js:47:14
    │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1975:25
    │ test@http://web-platform.test:8000/resources/testharness.js:535:30
    └ @http://web-platform.test:8000/FileAPI/url/url-format.any.js:39:5
    ▶ Unexpected subtest result in /FileAPI/url/sandboxed-iframe.html:
    │ FAIL [expected PASS] XHR should return Content-Type from Blob
    │   → assert_equals: expected (string) "image/png" but got (object) null
    │
    │ xhr.onloadend<@http://web-platform.test:8000/FileAPI/url/url-with-xhr.any.js:45:18
    │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1975:25
    └ Test.prototype.step_func_done/<@http://web-platform.test:8000/resources/testharness.js:2016:32
    ```

2.  /fetch/security/dangling-markup-mitigation.tentative.html
    ```
    ▶ Unexpected subtest result in /fetch/security/dangling-markup-mitigation.tentative.html:
    └ PASS [expected FAIL] <img id="dangling" src="/images/green-1x1.png?img=&lt;b">
    ▶ Unexpected subtest result in /fetch/security/dangling-markup-mitigation.tentative.html:
    └ PASS [expected FAIL] <img id="dangling" src="/images/green-1x1.png?img=&#10;b">
    ▶ Unexpected subtest result in /fetch/security/dangling-markup-mitigation.tentative.html:
    └ PASS [expected FAIL] <img id="dangling" src="/images/green-1x1.png?img=&amp;#10;b">
    ▶ Unexpected subtest result in /fetch/security/dangling-markup-mitigation.tentative.html:
    └ PASS [expected FAIL] <img id="dangling" src="/images/green-1x1.png?img=&amp;lt;b">
    ▶ Unexpected subtest result in /fetch/security/dangling-markup-mitigation.tentative.html:
    └ PASS [expected FAIL] <img id="dangling" src="/images/green-1x1.png?img=&amp;#10;b&amp;lt;c">
    ▶ Unexpected subtest result in /fetch/security/dangling-markup-mitigation.tentative.html:
    └ PASS [expected FAIL] \\n      <img id="dangling" src="\\n        /images/green-1x1.png?img=\\n      ">\\n
    ▶ Unexpected subtest result in /fetch/security/dangling-markup-mitigation.tentative.html:
    └ PASS [expected FAIL] \\n      <img id="dangling" src="\\n        /images/green-1x1.png?img=&amp;lt;\\n      ">\\n
    ▶ Unexpected subtest result in /fetch/security/dangling-markup-mitigation.tentative.html:
    └ PASS [expected FAIL] \\n      <img id="dangling" src="\\n        /images/green-1x1.png?img=&amp;#10;\\n      ">\\n
    ▶ Unexpected subtest result in /fetch/security/dangling-markup-mitigation.tentative.html:
    │ FAIL [expected PASS] <img id="dangling" src="/images/green-1x1.png?img=&#10;&lt;b">
    │   → assert_equals: Height expected 0 but got 1
    │ FAIL [expected PASS] <img id="dangling" src="/images/green-1x1.png?img=&lt;&#10;b">
    │   → assert_equals: Height expected 0 but got 1
    │ FAIL [expected PASS] \\n      <img id="dangling" src="/images/green-1x1.png?img=\\n        &lt;\\n        &#10;b\\n      ">\\n
    │   → assert_equals: Height expected 0 but got 1
    │
    │ assert_img_not_loaded/<@http://web-platform.test:8000/fetch/security/dangling-markup-mitigation.tentative.html:79:22
    │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1975:25
    └ Test.prototype.step_func_done/<@http://web-platform.test:8000/resources/testharness.js:2016:32
    ```


3.  /html/semantics/embedded-content/the-iframe-element/iframe_sandbox_anchor_download_allow_downloads.sub.tentative.html
    ```
    ▶ OK [expected TIMEOUT] /html/semantics/embedded-content/the-iframe-element/iframe_sandbox_anchor_download_allow_downloads.sub.tentative.html
    ▶ Unexpected subtest result in /html/semantics/embedded-content/the-iframe-element/iframe_sandbox_anchor_download_allow_downloads.sub.tentative.html:
    │ FAIL [expected TIMEOUT] <a download> triggered download in sandbox is allowed by allow-downloads.
    └   → assert_unreached: Unexpected navigation. Reached unreachable code
    ```

4.  /html/semantics/embedded-content/the-iframe-element/iframe_sandbox_anchor_download_block_downloads.sub.tentative.html
    ```
    ▶ Unexpected subtest result in /html/semantics/embedded-content/the-iframe-element/iframe_sandbox_anchor_download_block_downloads.sub.tentative.html:
    │ FAIL [expected PASS] <a download> triggered download in sandbox is blocked.
    └   → assert_unreached: Unexpected navigation. Reached unreachable code
    ```

5.  /html/semantics/embedded-content/the-iframe-element/iframe_sandbox_navigation_download_allow_downloads.sub.tentative.html
    ```
    ▶ OK [expected TIMEOUT] /html/semantics/embedded-content/the-iframe-element/iframe_sandbox_navigation_download_allow_downloads.sub.tentative.html
    ▶ Unexpected subtest result in /html/semantics/embedded-content/the-iframe-element/iframe_sandbox_navigation_download_allow_downloads.sub.tentative.html:
    │ FAIL [expected TIMEOUT] Navigation resulted download in sandbox is allowed by allow-downloads.
    └   → assert_unreached: Unexpected navigation. Reached unreachable code
    ```

6.  /html/semantics/embedded-content/the-iframe-element/iframe_sandbox_navigation_download_block_downloads.sub.tentative.html
    ```
    ▶ Unexpected subtest result in /html/semantics/embedded-content/the-iframe-element/iframe_sandbox_navigation_download_block_downloads.sub.tentative.html:
    │ FAIL [expected PASS] Navigation resulted download in sandbox is blocked.
    └   → assert_unreached: Unexpected navigation. Reached unreachable code
    ```

7.  /html/semantics/embedded-content/the-img-element/move-element-and-scroll.html
    ```
    ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/move-element-and-scroll.html:
    │ FAIL [expected PASS] Test that <img> below viewport is not loaded when moved to another document and then scrolled to
    └   → assert_unreached: The below viewport image should not load Reached unreachable code
    ```

8.  /html/semantics/embedded-content/the-img-element/sizes/sizes-dynamic-001.html,
    /html/semantics/embedded-content/the-img-element/sizes/sizes-dynamic-001-ref.html
    ```
    ▶ FAIL [expected PASS] /html/semantics/embedded-content/the-img-element/sizes/sizes-dynamic-001.html
    │   → /html/semantics/embedded-content/the-img-element/sizes/sizes-dynamic-001.html 7c4e7406993aa5c92015394610c118292c0593aa
    └   → /html/semantics/embedded-content/the-img-element/sizes/sizes-dynamic-001-ref.html df91e1e5441250d340a1203a86f926cde5c0426e
    ```

9. /referrer-policy/generic/sandboxed-iframe-with-opaque-origin.html
    ```
    ▶ OK [expected TIMEOUT] /referrer-policy/generic/sandboxed-iframe-with-opaque-origin.html
    ▶ Unexpected subtest result in /referrer-policy/generic/sandboxed-iframe-with-opaque-origin.html:
    └ PASS [expected TIMEOUT] Sandboxed iframe with opaque origin doesn't send referrers.
    ▶ Unexpected subtest result in /referrer-policy/generic/sandboxed-iframe-with-opaque-origin.html:
    │ FAIL [expected TIMEOUT] Sandboxed iframe with tuple origin sends referrers.
    │   → assert_equals: expected (string) "http://web-platform.test:8000/referrer-policy/generic/sandboxed-iframe-with-opaque-origin.html" but got (undefined) undefined
    │
    │ testSandboxedIframe/</<@http://web-platform.test:8000/referrer-policy/generic/sandboxed-iframe-with-opaque-origin.html:19:28
    │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1975:25
    └ Test.prototype.step_func/<@http://web-platform.test:8000/resources/testharness.js:2000:35
    ```

10. /html/semantics/document-metadata/the-base-element/base_srcdoc.html
    ```
    ▶ Unexpected subtest result in /html/semantics/document-metadata/the-base-element/base_srcdoc.html:
    └ PASS [expected FAIL] base element in srcdoc document should resolve against its fallback base URI
    ```

11. /css/css-values/viewport-units-after-font-load.html
    ```
    ▶ OK [expected TIMEOUT] /css/css-values/viewport-units-after-font-load.html
    ▶ Unexpected subtest result in /css/css-values/viewport-units-after-font-load.html:
    └ PASS [expected TIMEOUT] Viewport units are correctly updated after resize even if a font load has happened before
    ```

**Comment by jdm** [comment]
A bunch of the new failures that I looked at like https://searchfox.org/mozilla-central/source/testing/web-platform/tests/html/semantics/embedded-content/the-img-element/move-element-and-scroll.html make sense since they rely on features we don't support yet. Looks good!

**Comment by utsavoza** [comment]
@jdm, I looked at the test failures but I am not sure about what actually caused the failures in:
1. `/FileAPI/url/sandboxed-iframe.html`
2. `/fetch/security/dangling-markup-mitigation.tentative.html`
3. `/html/semantics/embedded-content/the-img-element/sizes/sizes-dynamic-001.html`

Also, I noticed there are many open issues that partly concern the test failures, but I am also not sure if we need to create any new issues to track any of these?

**Comment by jdm** [comment]
sandboxed-iframe.html and dangling-markup-mitigation.html started resolving URLs correctly inside the srcdoc iframe for the first time, so the test failures represent tests that didn't run at all previously.

I believe the failure in sizes-dynamic-001.html is caused again by the URLs now resolving correctly, and I think we don't implement intrinsic sizes for responsive images so the test now loads images that don't match the reference file.

**Comment by utsavoza** [comment]
I see. Let me know if we require any other changes.

**Pull request merged**